### PR TITLE
Distributed: Testing gather/allgather

### PR DESCRIPTION
**Pull request by jjhursey**
#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

`gather` and `allgather` are supported by Spyre Comms. Add unit tests to Torch Spyre for basic testing of the interface down to Spyre Comms via the `torch.distributed` interface. More extensive testing happens at the Spyre Comms level.

#### Which issue(s) this PR is related to:

https://github.com/torch-spyre/torch-spyre/issues/1902

### CODE REVIEW

File: tests/distributed/test_gather.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        """
+        Helper method to test gather with specific parameters.
+
+        Args:
+            shape: Tensor shape
+            dtype: Tensor data type
+            dst: Destination rank for gather
+        """
+        fill_value = self.comm_rank + 1
+        input_tensor = torch.full(shape, fill_value=fill_value, dtype=dtype)
```
Comment (ani300) [inline tests/distributed/test_gather.py]: could we make the tensor more varied in-rank to ensure ordering within ranks is also well preserved?

File: tests/distributed/test_gather.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        """Test gather to rank 0 with float16 tensors."""
+        self._test_gather_helper(shape=(128,), dtype=torch.float16, dst=0)
+
+    def test_gather_rank_zero_float32(self):
+        """Test gather to rank 0 with float32 tensors."""
+        self._test_gather_helper(shape=(256,), dtype=torch.float32, dst=0)
+
+    def test_gather_rank_zero_int32(self):
+        """Test gather to rank 0 with int32 tensors."""
+        self._test_gather_helper(shape=(192,), dtype=torch.int32, dst=0)
```
Comment (ani300) [inline tests/distributed/test_gather.py:181]: do we need to test non-rank 0 gathers?

File: tests/distributed/test_allgather.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            result = output_list[rank_idx].to("cpu")
+            expected = torch.full(shape, fill_value=rank_idx + 1, dtype=dtype)
+            self._assert_tensor_equal(
+                result,
+                expected,
+                dtype,
+                f"Rank {self.comm_rank}: allgather result incorrect for source rank {rank_idx}",
+            )
+
+    def test_allgather_float16(self):
```
Comment (ani300) [inline tests/distributed/test_allgather.py:163]: should we test with 2d tensors too?

File: tests/distributed/test_gather.py (hunk 1 above)
Comment (jjhursey) [reply tests/distributed/test_gather.py]: Are you thinking of a different fill pattern (instead of a single value per rank, make it a range of values per rank)? Or do you mean something different?

File: tests/distributed/test_gather.py (hunk 2 above)
Comment (jjhursey) [reply tests/distributed/test_gather.py:181]: Yeah we can add that

File: tests/distributed/test_allgather.py (hunk 3 above)
Comment (jjhursey) [reply tests/distributed/test_allgather.py:163]: Sure. Is 2D sufficient or do we need other dimensions?

**Comment by jjhursey** [comment]
Rebased to pick up https://github.com/torch-spyre/torch-spyre/pull/2833 CI fix

File: tests/distributed/test_allgather.py (hunk 3 above)
Comment (jjhursey) [reply tests/distributed/test_allgather.py:163]: Added 2D in latest update.

File: tests/distributed/test_gather.py (hunk 1 above)
Comment (ani300) [reply tests/distributed/test_gather.py]: I was thinking rank 0 gets 0-128, rank 1 get 256-384, rank 2 512-640, aka contiguous ranges in each rank but discontiguous across ranks, to ensure whatever we are doing makes sense both intra and across- ranks

File: tests/distributed/test_allgather.py (hunk 3 above)
Comment (ani300) [reply tests/distributed/test_allgather.py:163]: 2D should be enough yes, I was thinking of testing a layout that is truly tiled, as 1D does not tile by default

File: tests/distributed/test_gather.py (hunk 1 above)
Comment (jjhursey) [reply tests/distributed/test_gather.py]: Sounds good. I'll make that change.

Comment (jjhursey) [reply tests/distributed/test_gather.py]: done in latest push.

File: tests/distributed/test_gather.py (hunk 2 above)
Comment (jjhursey) [reply tests/distributed/test_gather.py:181]: In the latest push, but it did uncover a corner case that is being fixed in Spyre Comms. So the test will fail until the fix makes its way into the CI image.

**Review by ani300 — DISMISSED** [review]
tests lgtm at this point! thanks

**Comment by ani300** [comment]
@jjhursey it seems there is an error with non-zero rank gather?

**Comment by jjhursey** [comment]
> @jjhursey it seems there is an error with non-zero rank gather?

Yeah. I'm waiting for the CI image to pick up the fix merged today. Once it has that fix, then we can retest. With that fix everything passed clean in my testing.

**Comment by jjhursey** [comment]
CI Image has been updated and all checks are now passing. This is ready for final review.

**Review by ani300 — APPROVED** [review]

**Comment by mudhakar** [comment]
Minor comments:
1. _assert_tensor_equal ... duplicated code between test_allgather and test_gather....
2. self.assertTrue(False, error_msg)  --> had to read this twice.... self.fail() may be simpler.. similar for self.assertTrue(True, error_msg) --> this is just a pass with error_msg as a comment
3. in float16 exact int test:  rank * num_elements < 1024 is required, correct?
4. consider vectorizing: for i in range(num_elements): expected_tensor[i] = ... -> (torch.arange(num_elements) * comm_size + const).

**Comment by mudhakar** [comment]
@jjhursey ^^

**Review by mudhakar — APPROVED** [review]
lgtm.

For the comments above @flaviabeo to open issue and followup PR.

**Review by ashokponkumar — APPROVED** [review]
